utils/src/lib.rs

```diff
@@ -65,9 +72,13 @@ pub fn union(a: &[u8], b: &[u8]) -> Vec<u8> {
 }
 
 /// Concatenate a namespace and a message, prepended by a varint encoding of the namespace length.
+/// Method is the identity function if the namespace is empty.
 ///
 /// This produces a unique byte sequence (i.e. no collisions) for each `(namespace, msg)` pair.
 pub fn union_unique(namespace: &[u8], msg: &[u8]) -> Vec<u8> {
+    if namespace.is_empty() {
```

Comment:
@BrendanChou just making sure you see this tweak (given your prior work)

Reply:
This shortcut can result in a collision because then you can have the result of this function be an arbitrary vector of bytes by leaving the `namespace` empty and setting the `msg` to be whatever you want the output to be.

In order to make sure that each unique input results in a unique output, you would need to append a `0x00` byte to the output before `msg`. Otherwise, for example, if you leave this shortcut in, you have a collision like `union_unique({}, {0x01, 0x02, 0x03}) == union_unique({0x02}, {0x03})`

Reply:
I wonder if we should instead make namespace `Option<&[u8]>`?

Reply:
Can we just say that uniqueness is only guaranteed for non-empty namespaces?   Otherwise we could introduce a namespace-free variant to leverage these test vectors.

Reply:
I will do the Option<> approach in a second commit and we can see how it looks and make a call.